### PR TITLE
feat(ontology): add canonical Package node to ontology

### PR DESCRIPTION
**Pull request by jychp**
### Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring (no functional changes)
- [ ] Documentation update
- [ ] Other (please describe):

### Summary

Add a dedicated canonical `Package` ontology node (like `Device`/`User`/`PublicIP`) that aggregates software packages discovered across different scanning tools (Trivy, Syft).

The `Package:Ontology` node is deduplicated by `normalized_id` and linked to source nodes via `DETECTED_AS` relationships. Relationships from source nodes are propagated to the canonical node via `OntologyRelMapping` Cypher queries:

- **DEPLOYED**: `(:Package)-[:DEPLOYED]->(:ECRImage|GCP images|GitLabContainerImage)` via TrivyPackage
- **AFFECTS**: `(:TrivyImageFinding)-[:AFFECTS]->(:Package)` via TrivyPackage
- **DEPENDS_ON**: `(:Package)-[:DEPENDS_ON]->(:Package)` via SyftPackage dependency graph
- **SHOULD_UPDATE_TO**: `(:Package)-[:SHOULD_UPDATE_TO]->(:TrivyFix)` via TrivyPackage (compatibility bridge; deprecated for removal in v1)

This enables cross-tool querying like `MATCH (p:Package)-[:DEPLOYED]->(img) WHERE (f:TrivyImageFinding)-[:AFFECTS]->(p)` regardless of which scanner discovered the package.

### Related issues or links

- N/A

### Breaking changes

- `TrivyPackageSchema` primary label changed from `Package` to `TrivyPackage`. Existing Cypher queries using `MATCH (p:Package)` to find Trivy packages must be updated to either:
  - `MATCH (p:TrivyPackage)` for Trivy-specific queries
  - `MATCH (p:Package:Ontology)` for cross-tool canonical queries (requires ontology sync to run first)
- `TrivyFixToPackageRel` now targets `TrivyPackage` instead of `Package`

### How was this tested?

- Integration tests for ontology Package node creation, DETECTED_AS relationships, and all 7 propagated relationships (DEPLOYED x4, AFFECTS, DEPENDS_ON, SHOULD_UPDATE_TO compatibility bridge)
- All existing Trivy tests (8/8) updated and passing
- All Syft tests (3/3) passing unchanged
- All ontology tests (5/5) passing
- All unit tests (33/33) passing
- Lint clean (`make test_lint`)

#### General
- [x] I have read the [contributing guidelines](https://cartography-cncf.github.io/cartography/dev/developer-guide.html).
- [x] The linter passes locally (`make lint`).
- [x] I have added/updated tests that prove my fix is effective or my feature works.

#### Proof of functionality
- [x] New or updated unit/integration tests.

#### If you are changing a node or relationship
- [ ] Updated the [schema documentation](https://github.com/cartography-cncf/cartography/tree/master/docs/root/modules).
- [ ] Updated the [schema README](https://github.com/cartography-cncf/cartography/blob/master/docs/schema/README.md).

#### If you are implementing a new intel module
- [x] Used the NodeSchema [data model](https://cartography-cncf.github.io/cartography/dev/writing-intel-modules.html#defining-a-node).

### Notes for reviewers

- `SHOULD_UPDATE_TO` is currently propagated to canonical `Package` as a temporary compatibility bridge (`# DEPRECATED` in code), and is planned for removal in Cartography v1.
- GitHub `Dependency` is intentionally not mapped into canonical `Package` in this PR; follow-up tracked in #2406.
- No `config.ontology_packages_source` setting — all registered modules are used by default.

### CODE REVIEW

File: cartography/models/ontology/package.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    rel_label: str = "DETECTED_AS"
+    properties: PackageToNodeRelProperties = PackageToNodeRelProperties()
+
+
+@dataclass(frozen=True)
+class PackageSchema(CartographyNodeSchema):
+    label: str = "Package"
+    extra_node_labels: ExtraNodeLabels = ExtraNodeLabels(["Ontology"])
+    properties: PackageNodeProperties = PackageNodeProperties()
+    scoped_cleanup: bool = False
```
[the remark this answers is not in the document]
Comment (jychp) [reply cartography/models/ontology/package.py:136]: Package is an ontology node not linked to a tenant, that's the intended behavior (specific case)

File: cartography/models/trivy/fix.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ class TrivyFixToPackageRelProperties(CartographyRelProperties):
 
 @dataclass(frozen=True)
 class TrivyFixToPackageRel(CartographyRelSchema):
-    target_node_label: str = "Package"
+    target_node_label: str = "TrivyPackage"
```
[the remark this answers is not in the document]
Comment (jychp) [reply cartography/models/trivy/fix.py:30]: Documentation updated

File: cartography/intel/ontology/packages.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+
+@timeit
+def sync(
+    neo4j_session: neo4j.Session,
+    update_tag: int,
+    common_job_parameters: dict[str, Any],
+) -> None:
+    data = get_source_nodes_from_graph(neo4j_session, [], "packages")
+    load_packages(
```
[the remark this answers is not in the document]
Comment (jychp) [reply cartography/intel/ontology/packages.py:23]: This function queries the graph for nodes across multiple modules (Trivy, Syft, GitHub). In integration tests, only some modules are loaded the mock avoids having to set up the entire graph for each module.

File: tests/integration/cartography/intel/ontology/test_packages.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        MERGE (p2:SyftPackage {id: 'npm|body-parser|1.20.2'})
+        SET p2.normalized_id = 'npm|body-parser|1.20.2',
+            p2.name = 'body-parser', p2.version = '1.20.2',
+            p2.type = 'npm'
+        MERGE (p1)-[:DEPENDS_ON]->(p2)
+        """,
+    )
+
+
+@patch.object(
```
[the remark this answers is not in the document]
Comment (jychp) [reply tests/integration/cartography/intel/ontology/test_packages.py:62]: This function queries the graph for nodes across multiple modules (Trivy, Syft, GitHub). In integration tests, only some modules are loaded the mock avoids having to set up the entire graph for each module.

File: docs/root/modules/syft/schema.md [hunk 5]
```diff
@@ -1,70 +1,55 @@
...
 
-Self-referential dependency relationships between SyftPackage nodes.
+- A SyftPackage depends on another SyftPackage.
 
-```
-(:SyftPackage)-[:DEPENDS_ON]->(:SyftPackage)
-```
+    ```
+    (SyftPackage)-[:DEPENDS_ON]->(SyftPackage)
+    ```
```
[the remark this answers is not in the document]
Comment (kunaals) [reply docs/root/modules/syft/schema.md:26]: eh its okay

File: cartography/models/ontology/mapping/__init__.py [hunk 6]
```diff
@@ -54,6 +56,7 @@
 ONTOLOGY_NODES_MAPPING: dict[str, dict[str, OntologyMapping]] = {
     "users": USERS_ONTOLOGY_MAPPING,
     "devices": DEVICES_ONTOLOGY_MAPPING,
+    "packages": PACKAGES_ONTOLOGY_MAPPING,
```
[the remark this answers is not in the document]
Comment (kunaals) [reply cartography/models/ontology/mapping/__init__.py:59]: this is there

File: cartography/intel/github/repos.py [hunk 7]
```diff
@@ -858,6 +927,23 @@ def _transform_dependency_graph(
             # Create manifest ID for the HAS_DEP relationship
             manifest_id = f"{repo_url}#{manifest_path}"
 
+            # Compute ontology fields for Package matching
+            exact_version = _extract_exact_version(requirements)
+            purl_type = GITHUB_ECOSYSTEM_TO_PURL_TYPE.get(package_manager)
+            dep_purl: Optional[str] = None
+            dep_normalized_id: Optional[str] = None
+            dep_type: Optional[str] = None
```
Comment (kunaals) [inline cartography/intel/github/repos.py]: can we include `packageUrl` in the graphql query and prefer it here when building `purl`/`normalized_id`? right now we reconstruct from `requirements` + `packageManager`, which is lossy for several ecosystems and can cause false negatives in ontology matching. github added `DependencyGraphDependency.packageUrl` for purl-based ecosystems (april 2025): https://github.blog/changelog/2025-04-03-dependency-graph-supports-all-purl-identified-package-ecosystems/

File: cartography/models/trivy/package.py [hunk 8]
```diff
@@ -98,9 +97,8 @@ class TrivyPackageToFindingRel(CartographyRelSchema):
 
 @dataclass(frozen=True)
 class TrivyPackageSchema(CartographyNodeSchema):
-    label: str = "Package"
```
Comment (kunaals) [inline cartography/models/trivy/package.py:101]: this label rename makes sense, but can we add an explicit migration/cleanup step for legacy `:Package` nodes from pre-change runs? otherwise old trivy package nodes can still match `MATCH (p:Package)` and contaminate ontology package queries after upgrade.

File: cartography/intel/github/repos.py [hunk 9]
```diff
@@ -803,6 +804,74 @@ def _transform_dependency_manifests(
         logger.info(f"Found {manifests_added} dependency manifests in {repo_name}")
 
 
+GITHUB_ECOSYSTEM_TO_PURL_TYPE: Dict[str, str] = {
+    "NPM": "npm",
+    "PIP": "pypi",
+    "MAVEN": "maven",
+    "NUGET": "nuget",
+    "RUBYGEMS": "gem",
+    "GO": "golang",
```
Comment (kunaals) [inline cartography/intel/github/repos.py]: small but important: github dependency graph ecosystem uses `RUST` (not `CARGO`) in graphql. can we add `"RUST": "cargo"` (or equivalent handling) so rust deps get purl/normalized_id?

File: cartography/intel/github/repos.py [hunk 10]
```diff
@@ -803,6 +804,74 @@ def _transform_dependency_manifests(
...
+
+    Returns the version string for exact pins (e.g. "18.2.0", "= 4.2.0", "== 4.2.0").
+    Returns None for ranges, wildcards, or empty strings.
+    """
+    if not requirements:
+        return None
+    req = requirements.strip()
+    if not req:
+        return None
+    # Match "= version" or "== version" prefix
```
Comment (kunaals) [inline cartography/intel/github/repos.py]: can we tighten this parser so it only accepts true exact pins? right now malformed strings like `===1.0.0` can slip into weird outputs. i’d suggest rejecting `===`, rejecting extra-token forms, and adding unit tests for these edge cases.

File: cartography/models/ontology/package.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    # ontology-derived edges by lastupdated on each sync run.
+    other_relationships: OtherRelationships = OtherRelationships(
+        rels=[
+            PackageToTrivyPackageRel(),
+            PackageToSyftPackageRel(),
+            PackageToECRImageRel(),
+            PackageToGCPArtifactRegistryContainerImageRel(),
+            PackageToGCPArtifactRegistryPlatformImageRel(),
+            PackageToGitLabContainerImageRel(),
+            PackageToTrivyFixRel(),
```
[the remark this answers is not in the document]
Comment (kunaals) [reply cartography/models/ontology/package.py:147]: edited
